### ISSUE TITLE
Support login SAML

### ISSUE THREAD

**Issue by NgoHuy** [adrienverge/openfortivpn#867]
Hi,
I see that openfortivpn supported 2FA but no supports SAML. Can you implement this feature?

Best regards,
Severus

**Comment by tjsd** [comment]
Any update on this from dev? Is it hard to implement?

**Comment by DimitriPapadopoulos** [comment]
While I probably won't have time myself, pull requests are welcome.

Should we use a library that implements SAML? It looks [Lasso](https://lasso.entrouvert.org/) is or used to be the de facto standard for C programs, but as far as I can understand it's for writing service providers (SP) such as a FortiGate, not clients.

**Comment by DimitriPapadopoulos** [comment]
Some FortiGate documentation:
[Technical Tip: Configuring SAML SSO login for FortiGate administrators with Azure AD acting as SAML IdP](https://kb.fortinet.com/kb/documentLink.do?externalID=FD45699)

**Comment by DimitriPapadopoulos** [comment]
This might be of interest: [gp-saml-gui](https://github.com/dlenski/gp-saml-gui)

It would help to have screen captures of what SAML authentication looks like with FortiCLient.

**Comment by DimitriPapadopoulos** [comment]
My guess is that the cookie is retrieved by a possibly independent process, then fed to the VPN client which does not need to re-authenticate, just use the cookie. If so, the question is, how do we retrieve the cookie? Can you describe the FortiClient workflow (with screen captures) or at least the expected workflow?

**Comment by tjsd** [comment]
I can provide some screenshots:
![image](https://user-images.githubusercontent.com/7138125/114383547-f3541500-9b8d-11eb-9115-51904798d35b.png)
![image](https://user-images.githubusercontent.com/7138125/114383613-07981200-9b8e-11eb-99b5-c7155eb12c6b.png)

You get a popup that is asking for you microsoft credentials:
![image](https://user-images.githubusercontent.com/7138125/114383701-25657700-9b8e-11eb-9560-3a3718a26863.png)

**Comment by memco** [comment]
Hi, interested in helping to get this implemented. I can help with screenshots and I can see if I can provide some logs or run a sample session with mitmproxy to get request & responses (sanitized ofc). What kind of info would be helpful?

**Comment by heimmdal** [comment]
I am also willing to help with this feature. Official client supports it. Would be wonderful if openfortivpn does it too.

**Comment by DimitriPapadopoulos** [comment]
Patches are welcome 😀

**Comment by DimitriPapadopoulos** [comment]
Perhaps you can have a look at [_openconnect_](https://gitlab.com/openconnect/openconnect). I am not certain it supports SAML yet, but it might be worth trying. If it does not, I suggest you open an issue there too. Perhaps they will have the resources to implement it more rapidly.

**Comment by NgoHuy** [comment]
> Perhaps you can have a look at [_openconnect_](https://gitlab.com/openconnect/openconnect). I am not certain it supports SAML yet, but it might be worth trying. If it does not, I suggest you open an issue there too. Perhaps they will have the resources to implement it more rapidly.

openconnet does not support SAML but other project like https://github.com/yuezk/GlobalProtect-openconnect did it.

**Comment by DimitriPapadopoulos** [comment]
I think they have a branch in the works - not the _master_ branch. Perhaps support for SAML will be available after _openconnect_ 9.0 is released.

Honestly, at this point and for lack of resources, I probably won't be able to write anything for _openfortivpn_ before they do...

**Comment by msaunois-dzr** [comment]
Hi,

The company I work for will enable the SAML authentication. I'm not a C developer and thus I can not really help to implement the feature. But if you need some help for testing I would be glad to help you!

The bad thing is that I tested the official client on Gnome and KDE (Fedora and Arch) to help my IT team to validate the client on Linux, and it works fine, but I use i3 and I did not manage to get a working connection with it. When I try to connect the client opens the authentication form of the SSO platform used in my company (Okta), I fill my credentials but the client never completes the auth process. Of course nothing in my logs or in the output when I run the app from my terminal... I thought about some polkit or PAM issues but I could not find anything.

Does somebody have some tricks about that to help me waiting for a SAML implementation on openfortivpn ?

PS: Thank you very much for creating openfortivpn, it allowed me to not use the bloated and buggy official client for so long!

**Comment by lart2150** [comment]
I don't code in C but I might be able to fill in some of the gaps in case someone wants to take this up.  I enabled ssl debug logging on a fortigate a connected with the windows client and I saw the following requests.  I don't think you need any kind of saml library just a browser and a way to extract part of the response.  The official client uses electronjs to display the login.

1. /remote/info
- contains remoteauthtimeout in seconds.  this while not needed is important as the user can take a long time to login but the fortigate will timeout the session.
2. /remote/saml/start
- redirects to the saml idp login page
3. /remote/saml/login
- redirected here from the IDP.  a cookie gets set here
4. /sslvpn/portal.html
5. /remote/fortisslvpn
- looks like it sets a cookie via javascript
`<script>
var cookie_str='pc2QKF5dD...8pgsU9zoYJbwP0H8=';
</script>`
6. /remote/licensecheck
- I don't know if this is required.  I don't have ems but 🤷‍♂️
7. /remote/fortisslvpn_xml
- this looks like the list of routes/dns/etc
8. /remote/sslvpn-tunnel?uuid=[HEX]

**Comment by rtsantanna** [comment]
I'm not a developer either, but I can help test any patch or new version.

The company enabled yesterday, of course, without testing ignoring Linux users (they only support those who use Windows), and in my case with Arch, not even the official client (version 7.0.0.18) is working, but a colleague with Ubuntu didn't work either .

**Comment by rtsantanna** [comment]
I did a clean install, to test the crap of the manufacturer's official client.

The packages you installed from installing the forticlient package were:

Packages to install (from repository https://repo.fortinet.com/repo/7.0/centos/8/os/x86_64): forticlient

Below, the installed dependencies:
libXScrnSaver
m4
mailx
make
ncurses-compat-libs
nss-tools
patch
postfix
redhat-lsb-core
redhat-lsb-submod-security
spax
yum-utils


After that, I called the binary from a terminal, to log what would be done. Configured the VPN portal (just put the address and the SAML option).

Below, I put the output of the process (I decharacterized so I don't have problems):

`[infra@teste-fortinet ~]$ forticlient 
Platform detected: fedora
[ '/opt/forticlient/gui/FortiClient-linux-x64/FortiClient' ]
Saml - init
did-finish-load
renderer ready -  IPC_RENDERER_REQUEST.FETCH_INVITATION_CODE 2021-08-19T20:04:02.768Z
this.inviteCode=null
IPC_RENDERER_REQUEST.LOADED
Events - processArgv
[ '/opt/forticlient/gui/FortiClient-linux-x64/FortiClient' ]
certificate-error url=https://vpn.address.com/remote/info
saml-login https://vpn.address.com:443/remote/saml/start
Saml - doSamlAuth
{ authTimeout: '600',
  ignoreCert: false,
  type: 1,
  url: 'https://vpn.address.com:443/remote/saml/start' }
Saml - loadUrlWithType type=1 url=https://vpn.address.com:443/remote/saml/start
samlReq.url = https://vpn.address.com:443/remote/saml/start
samlReq.type = 1
beginUrl.href = https://vpn.address.com/remote/saml/start
certificate-error url=https://vpn.address.com/remote/saml/start
Gtk-Message: 17:05:22.632: GtkDialog mapped without a transient parent. This is discouraged.
continue-error-certificate https://vpn.address.com/gateway-realm:443/remote/saml/start
Saml - reloadWindow authTimeout = 600
Saml - loadUrlWithType type=1 url=https://vpn.address.com/gateway-realm:443/remote/saml/start
certificate-error url=https://vpn.address.com/gateway-realm:443/remote/saml/start
FortiSAMLInstance.IgnoreCertificate url=https://vpn.address.com/gateway-realm:443/remote/saml/start
samlType=1 logDomainCookies - https://vpn.address.com
certificate-error url=https://vpn.address.com/sslvpn/css/ssl_style.css
FortiSAMLInstance.IgnoreCertificate url=https://vpn.address.com/sslvpn/css/ssl_style.css
samlType=1 logDomainCookies - https://vpn.address.com
certificate-error url=https://vpn.address.com/sslvpn/css/ssl_style.css
FortiSAMLInstance.IgnoreCertificate url=https://vpn.address.com/sslvpn/css/ssl_style.css
samlType=1 logDomainCookies - https://vpn.address.com
Saml - 'did-finish-load'
Saml - currentPath=https://vpn.address.com/gateway-realm:443/remote/saml/start
samlType=1 logDomainCookies - https://vpn.address.com
SAML - current path: https://vpn.address.com/gateway-realm:443/remote/saml/start
Saml - handleRedirect url=https://vpn.address.com/gateway-realm:443/remote/saml/start
Saml will-nativgate newUrl=https://vpn.address.com/remote/login
Saml - handleRedirect url=https://vpn.address.com/remote/login
Saml will-redirect url=https://vpn.address.com/remote/login?lang=en
certificate-error url=https://vpn.address.com/sslvpn/js/login.js?q=17551d5deba7dad25bffcfacc62cb666
FortiSAMLInstance.IgnoreCertificate url=https://vpn.address.com/sslvpn/js/login.js?q=17551d5deba7dad25bffcfacc62cb666
samlType=1 logDomainCookies - https://vpn.address.com
certificate-error url=https://vpn.address.com/remote/fgt_lang?lang=en
FortiSAMLInstance.IgnoreCertificate url=https://vpn.address.com/remote/fgt_lang?lang=en
samlType=1 logDomainCookies - https://vpn.address.com
certificate-error url=https://vpn.address.com/remote/fgt_lang?lang=en
FortiSAMLInstance.IgnoreCertificate url=https://vpn.address.com/remote/fgt_lang?lang=en
samlType=1 logDomainCookies - https://vpn.address.com
Saml - 'did-finish-load'
Saml - currentPath=https://vpn.address.com/remote/login?lang=en
samlType=1 logDomainCookies - https://vpn.address.com
SAML - current path: https://vpn.address.com/remote/login?lang=en
Saml - handleRedirect url=https://vpn.address.com/remote/login?lang=en
Saml - 'close' state=4
cancelAndExit state=4 currentWindow=[object BrowserWindow]
closeServer
Saml - 'closed' state=4
`

**Comment by andyrowlett** [comment]
Can't offer any help, but also affected by the SAML issue. Bought a raspberry pi last year to use as a thin client, and been unable to do so since work added the SAML requirement (again, like other posters, non-windows is not really considered) - as far as I can tell, openfortivpn is the only way to use fortinet vpn on an arm device - so really hope this is possible!

**Comment by DimitriPapadopoulos** [comment]
For what it's worth, the `@master` branch of [OpenConnect](https://gitlab.com/openconnect/openconnect) does support Fortinet SSL VPNs, but I think it doesn't support SAML yet, either.

**Comment by jng3555** [comment]
Not sure of anybody is working on this. Using 1.17.1, instead of "/remote/login" - it goes to "/remote/saml/start". However, the redirect to "https://login.microsoftonline.com/" this part, I have no idea how to handle it.
Any suggestions are welcome.
Thanks.

**Comment by bowensong** [comment]
> Not sure of anybody is working on this. Using 1.17.1, instead of "/remote/login" - it goes to "/remote/saml/start". However, the redirect to "https://login.microsoftonline.com/" this part, I have no idea how to handle it. Any suggestions are welcome. Thanks.

I can see a few solutions, each has their pros and cons.

1. Show the URL to the user, ask the user to proceed in a real web browser, and eventually copy & paste the result back. The [Linux OneOrive client](https://github.com/skilion/onedrive) does this. It works, but I personally do not like the solution, because it requires extra user interactions.
2. Run a headless web browser and performs the necessary steps inside the browser. This allows non-interactive use, but may break if Microsoft changes their login page. Some people may also have a distaste of the added overhead of a headless web browser.
3. Dealing with the HTTP GET/POST requests in the code, and parse the HTML / JavaScript responses. This is the lightweight and non-interactive solution, but does require a bit of extra effort and may break if Microsoft changes their login page.
4. Let an external program handles it, pass the URL to the external program and expect the result back from it. User may choose to use any of the above methods to implement the external program, everyone is happy.

**Comment by debalance** [comment]
I expect there are multiple different SAML providers following a certain standard, I had to use Forticlient recently with SAML, there was no trace of anything Microsoft related.

**Comment by bowensong** [comment]
> I expect there are multiple different SAML providers following a certain standard, I had to use Forticlient recently with SAML, there was no trace of anything Microsoft related.

Hmm, in that case, I can only see two ways out of this - interactive authentication in a web browser or an external program.

I personally prefer the later with some reference implementations, and one of them is the interactive authentication. The user should be able to use the reference implementation out of box if they prefer, or implement their own non-interactive version for their SAML provider.

**Comment by gm-vm** [comment]
My company has just started requiring SAML. After battling a bit with the official Forticlient, which keeps dropping the connection soon after creating the tunnel, I decided to look into ways to get the VPN working with `openfortivpn`.

This is the result: https://github.com/gm-vm/openfortivpn/tree/saml

The solution is far from ideal, it requires you to login from a browser, grab the `SVPNCOOKIE` and paste it in the console, but it's good enough for me for now.

**Comment by MartialGeek** [comment]
Thank you @gm-vm!

I don't know if I made a mistake but I tried to add `saml = 1` in the config file and the program returns `WARN:   Bad key in configuration file: "saml".`

However using the flag `--saml` works fine, I just had to take the port into account for the login URL because my IT team set up another port.

**Comment by gm-vm** [comment]
@MartialGeek thank you for the fixes.

Meanwhile I wrote a tiny Qt application to grab `SVPNCOOKIE` more easily. It is basically a `QtWebEngine` that open the SAML page and prints the `SVPNCOOKIE` to stdout as soon as it is set.
This is the application: https://github.com/gm-vm/openfortivpn-webview

My idea was that of passing the path to this program to `openfortivpn` through some command line argument and have it run it as needed. It was too late when I remembered that `openfortivpn` must run as root and I definitely do not want to run a browser as root. I could get that to work the way I wanted with a combination of `sudo --preserve-env`, `fork()` and `setuid()`, but I did not really like how it ended up looking.

I therefore decided to change approach and have `SVPNCOOKIE` passed as argument. I have also made it possible to pass it through stdin, so that I can do something like this:
```
openfortivpn-webview my_host 2>/dev/null | sudo openfortivpn my_host --svp-cookie -
```

I have created a new branch here: https://github.com/gm-vm/openfortivpn/tree/svpn_cookie


As I mentioned in my previous message, I really doubt I can get this working reliably without a browser.

This is for example an overview of the steps needed when using Duo, which omits the required steps for the setup of the mobile application: https://duo.com/docs/sso#saml-identity-provider

**Comment by DimitriPapadopoulos** [comment]
I think it is OK to use a browser as a first step. It is also OK to launch the browser before openfortivpn as a first step.

Note the ongoing effort to implement SAML in [OpenConnect](https://gitlab.com/openconnect) too. You might want to have a look, see for example https://gitlab.com/search?group_id=500944&project_id=2335175&scope=issues&search=SAML.

**Comment by bowensong** [comment]
If the browser window is NOT lunched by the VPN software, how is reconnecting going to be handled?

Keep in mind the SVPNCOOKIE cookie may have fairly short expiry time or is limited to single use only, but other cookies for keeping track of the user login status may have much longer validity and can be used to get the short lived cookie again when needed.

For example, with the FortiClient VPN, it pops up the login window when reconnecting, and closes it in a few seconds without requiring any user interaction if the user is already logged in. The whole process is just a lots of HTTP redirection with the "set-cookie" headers.

**Comment by DimitriPapadopoulos** [comment]
Automatic reconnection with `--persistent` doesn't work well anyway, it resends the same password, which of course cannot work with SAML. I was against this option, the same can be achieved with an encapsulating script or proper init/systemd parameters, or by the NetworkManager plugin.

The proper way would indeed be to attempt first to reconnect with a cookie first, then re-enter new credentials. But you're right that reconnection with a cookie doesn't seem to work that well in Fortinet products anyway, see discussion about this somewhere in the [OpenConnect](https://gitlab.com/openconnect/openconnect) issues or merge requests.

**Comment by MartialGeek** [comment]
> I definitely do not want to run a browser as root
I think it is a wise decision!

> If the browser window is NOT lunched by the VPN software, how is reconnecting going to be handled?
We may need to watch the status of openfortivpn?
I mean if the program that runs the browser acts as a daemon, it could monitor the openfortivpn process and re-lunch a login window under certain circumstance (some status code for example)?

**Comment by DimitriPapadopoulos** [comment]
While the cookies have no expiration date, I seem to recall they cannot be re-used (depending on the configuration of the Fortigate?), or perhaps re-used only once (whatever the Fortigate parameters?).

What would currently need to be merged into openfortivpn to enable SAML in conjunction with [openfortivpn-webview](https://github.com/gm-vm/openfortivpn-webview)?

**Comment by gm-vm** [comment]
@DimitriPapadopoulos All the changes I have made all do pretty much the same thing, albeit in different ways: let me inject a `SVPNCOOKIE` rather than having `openvpnclient` retrieve it autonomously. I have been experimenting with different ways to do it, not sure yet which is the best way to do it.

`openfortivpn-webview` does nothing more than printing the `SVPNCOOKIE` returned by the server. It never changes here, as far as I can, and I can have both the browser and `openfortivpn` run concurrently with the same cookie. I could be just lucky with how the Fortigate has been configured, I honestly don't know.

Once I stop `openfortivpn` I cannot reuse the same cookie, but I think that's probably because it logs me out, which would be equivalent to logging out from the webpage.

**Comment by bowensong** [comment]
> @bowensong I just published the changes that I did to have `openfortivpn` launch the application: [gm-vm@bd81728](https://github.com/gm-vm/openfortivpn/commit/bd817285502fa12d7dc39d81501267313d032a8f) I think I did not even test it properly, it's probably buggy. Launching the browser in that way did not look to good to me (the code does also some wrong assumptions that could be fixed, such as root always having `/root` as `$HOME`), which is why I gave up on that solution.

TBH, you are overly complicating things. It's hard to gain root access from a non-privileged user, which usually means you are going to need use `sudo` or something similar. However, to drop root access is very imple, all you need to do is `setgid()` and `setuid()`. Because no new process is created, memory content and file descriptors are also preserved, you don't even need environment variables to pass things around.

In my opinion, the whole process is suppose to be as simple as `pipe()`, `fork()`, then in the child process `setgid()`, `setuid()` and launch the browser. The child and parent processes can communicate over the pipe. If bi-directional communication is desired, call `pipe()` twice before forking. To reconnect, the parent process (which is still running as root) can repeat the above process as necessary.

**Comment by gm-vm** [comment]
@bowensong what you describe is pretty much what I implemented in the commit I linked.

**Comment by bowensong** [comment]
> @bowensong what you describe is pretty much what I implemented in the commit I linked.

oops, I didn't read the code. I only saw the `sudo --preserve-env` part in your comment, and had guessed you were running `sudo -u USERNAME --preserve-env the-gui-program`. If the only thing you needed from the environment variables is the username/UID from `sudo`, I would say it may be better to ask the user to provide it to openfortivpn directly. This works even if openfortivpn was started by something else, such as a systemd service.

**Comment by gm-vm** [comment]
Yes, I'm relying on `sudo` to retrieve the original user as well as have the correct `$HOME` set and a bunch of `$XDG_` env variables. Maybe the application can work without the `$XDG_` variables, but `QtWebEngine` was complaining about them not being there. Nothing that cannot be provided manually though.

**Comment by bowensong** [comment]
> Yes, I'm relying on `sudo` to retrieve the original user as well as have the correct `$HOME` set and a bunch of `$XDG_` env variables. Maybe the application can work without the `$XDG_` variables, but `QtWebEngine` was complaining about them not being there. Nothing that cannot be provided manually though.

I think only UID is truly required, and should be provided by the user. The home directory (and group ID) can be retrieved with `getpwent()` if needed. I don't know what `$XDG_*` variables are needed, but I doubt they are a hard requirement for running a browser window. Additionally, the `$DISPLAY` environment variable is required to run a GUI application, and I personally would prefer to leave that to a wrapper shell script to deal with before handing over to the real browser. This way, the user doesn't even need to set `$DISPLAY` if they don't need a real browser.

**Comment by bowensong** [comment]
Actually, I thought about this again, why do we even bother to drop the root privilege in the openfortivpn? As long as the SAML handler is an external command, the user can always put `sudo -u USERNAME real-saml-handler --whatever-parameter` in it. openfortivpn can provide the URL to the handler via stdin, and get the cookies back from the stdout. Make the handler itself responsible for dropping root if and when desired. The reason for this is because not everyone will run a browser to do the SAML handling, and in some cases it may be acceptable or even desirable to run it as root.

**Comment by gm-vm** [comment]
Yes, that makes sense and it simplifies the code quite a bit. I've been using the other approach without issues in the last few days, but having `openfortivpn` launch the handler on its own works equally well for my usecase.

I've updated my `saml_handler` branch, there are maybe some more things that I need to clean up a bit, but maybe I can open a PR here to get some feedback.

**Comment by iamjla** [comment]
I like the approach with the cookie parameter (with maybe another parameter which returns the saml url). But as i was unable to compile a non segfaulting client from gm-vm's source i was unable to test it.
I'd love to see a public packaged version so i also put some cash up for grabs on bountysource if that is not a no go(maybe some will join).

**Comment by gm-vm** [comment]
@iamjla is the segfault caused by my changes or something else? (e.g., if you build from master it does not happen)

Also, did you try to build from the branch I used for the PR? If not, could you try that?

Maybe I missed some code path that depends on some configuration parameter.

**Comment by iamjla** [comment]
I've written a quick and ugly wrapper curl script to get the svpn cookie from our MFA provider and parse it to the openfortivpn cli. My colleagues are using it with both cookie branches(saml_cookie, svpn_cookie) without issues. I myself haven't had any hiccups either so i'd really consider the requested feature implemented. 
As there are too many MFA providers out there which handle the requests differently i'd think you'd need an extra repository with either a truckload of individual scripts for azure, okta, keycloak etc. as well as otp tools like yubikey, keypass or the webview as mentioned above to get the cookie for the client. 
I'll happily wait for a merge so my non tech savvy colleagues can download the package :)

**Comment by joso9001** [comment]
I tried out #974 yesterday by snagging the cookie manually from my browser. Even if the user interaction parts aren't nailed down yet it would be nice to have something like this merged as a first step.

**Comment by JockeTF** [comment]
I tried out #974 yesterday by manually snagging the cookie from my browser. It would be nice to have something like this merged even if the user interaction parts aren't nailed down yet.

**Comment by JockeTF** [comment]
I'm in the process of [adding a cookie field to NetworkManager-fortivpnssl](https://github.com/NetworkManager/NetworkManager-fortisslvpn/compare/main...JockeTF:cookie_saml). It currently requires specifying any password, but seems to work other than that. Just remember to specify the cookie on the form `SVPNCOOKIE={VALUE}`. It's a bit of a hack, but does the job if you want NetworkManager integration.

**Comment by gm-vm** [comment]
I did not bother to look into ways to integrate this with NetworkManager, especially because I'm finding my current setup (which I've just described briefly [here](https://github.com/gm-vm/openfortivpn-webview/wiki)) pretty convenient for my usecase. It is not as user-friendly as NetworkManager, but it's good enough for me.

Ideally NetworkManager would take care of launching a browser to let you login and grab the cookie, but that requires quite some work. Doing everything with curl is either not feasible or prone to frequent breakages.

Plus, I'm a bit worried about potential issues that may arise from not using full-fledged browser.

For example, Cisco Duo has stopped working with `QtWebEngine` all of the sudden. I had to resort to Electron to fix things, which I believe is what the official client uses. Interestingly, I did not have issues with a python script that launches GTK's WebKit, nor with Epiphany, which probably uses the same engine. I did have issues with Konqueror, which is yet another engine, so who knows what's going on.

**Comment by gm-vm** [comment]
@DimitriPapadopoulos I guess you meant [saml_handler](https://github.com/gm-vm/openfortivpn/tree/saml_handler), because `saml` is pretty much what has been merged already.

I simply thought that being able to pass session cookies when launching the program was enough and a lot simpler than having it launch an external application. A shell script is enough to do what `saml_handler` does.

However, I can create a pull request for that too if you (and by that I mean all reading this) think it's a good idea. I'm pretty sure @bowensong found that useful.

**Comment by bowensong** [comment]
TBH, I've found my solution with [openconnect](https://gitlab.com/openconnect/openconnect) and its `--cookie` option. It works like a charm.

Basically, in openconnect:
1. start with the user script performs the SAML authentication and retrieves the SVPNCOOKIE cookie
2. the user script runs `exec openconnect --protocol=fortinet --cookie=SVPNCOOKIE=...`
3. openconnect takes over and deals with everything else

In openfortivpn with the purposed change:
1. start with `openfortivpn --saml --saml-handler=...`
2. openfortivpn runs the user script
3. the user script performs the SAML authentication and retrieves the SVPNCOOKIE cookie
5. openfortivpn get the result from the user script, and continues

Needless to say, the openconnect's approach looks simpler. The only downside is openconnect cannot reconnect if the connection is lost, because the cookie value cannot be reused. I addressed that issue by restarting the entire user script if and when the process unexpectedly exits, and it has worked very well for me.

**Comment by DimitriPapadopoulos** [comment]
I must admit I hadn't looked into the details of that branch. If the `--cookie` option is enough, and the rest can implemented in a script, I am all for keeping the rest in a script.

If so, can we close this issue?

BTW, I seem to recall that OpenConnect attempts to re-use the cookie and, if it doesn't work, attempts to re-use the password. Isn't that the case? You might want to open an issue there if it isn't. Note that FortiGate devices might not support using the same cookie twice (it depends on the configuration of the Fortigate) and as far as I know, does not support using the same cookie thrice.

**Comment by gm-vm** [comment]
Yes, I think the newly added `--cookie` option is enough and this issue can be closed.

**Comment by bowensong** [comment]
AFAIK, openconnect does attempt to reuse the cookie. The password part is irrelevant, because it doesn't support password in SAML.

It may be related to the configuration on the fortigate server side as you said. In my case, the cookie is single use. As soon as the first connection is authenticated with the cookie, the cookie becomes invalid and cannot be reused.

Anyway, the `--cookie` option works fine for me. I'm happy for this to be closed once the `--cookie` option is added.

**Comment by kkuhle** [comment]
> 1. start with the user script performs the SAML authentication and retrieves the SVPNCOOKIE cookie
> 2. the user script runs `exec openconnect --protocol=fortinet --cookie=SVPNCOOKIE=...`

@bowensong will you elaborate on the "user script" to do this?  My company just started using SAML and I believe *openconnect* will start failing for me soon since I'm not using the cookie method you described.

**Comment by bowensong** [comment]
@kkuhle In short, it's a Python script with [Selenium](https://www.selenium.dev/) WebDriver. It automates the login process in a headless web browser, and then retrieve the cookie from the browser before passing it to openconnect.

**Comment by chung1905** [comment]
Thank you for `--cookie` option.

It works for me with Microsoft login.

Full instruction is:
- Go to `{host}:{port}/remote/saml/start?realm=foo`
- Login
- Copy cookie with name: `SVPNCOOKIE`
- Run the follow command
```
sudo openfortivpn {host}:{port} --cookie="SVPNCOOKIE={cookie}" --realm=foo
```

**Comment by DimitriPapadopoulos** [comment]
Just wondering, would it help to have  the `--cookie=` option wrap its argument with `SVPNCOOKIE={`...`}`, so that you can write as `--cookie="cookie"` instead of `--cookie="SVPNCOOKIE={cookie}"`?

**Comment by joso9001** [comment]
I think wrapping it automatically would make sense, or perhaps add some auto-detection if it needs to be added. I was surprised by having to add SVPNCOOKIE myself too. I spent a bunch of time being confused over why it didn't work.

**Comment by uishon** [comment]
> @kkuhle In short, it's a Python script with [Selenium](https://www.selenium.dev/) WebDriver. It automates the login process in a headless web browser, and then retrieve the cookie from the browser before passing it to openconnect.

See [here](https://gist.github.com/uishon/412205dfd8b3f48bf07a77e3cdd6948b), using python with selenium 4

**Comment by DimitriPapadopoulos** [comment]
> I think wrapping it automatically would make sense, or perhaps add some auto-detection if it needs to be added. I was surprised by having to add SVPNCOOKIE myself too. I spent a bunch of time being confused over why it didn't work.

Perhaps we should add `SVPNCOOKIE={` when it is missing.

**Comment by gm-vm** [comment]
> Perhaps we should add SVPNCOOKIE={ when it is missing.

I've just opened https://github.com/adrienverge/openfortivpn/pull/1035.

**Comment by LorenzISR** [comment]
I've just opened #1034 

It would be great if someone could test this and give me feedback.

**Comment by gm-vm** [comment]
> I've just opened https://github.com/adrienverge/openfortivpn/pull/1034
>
> It would be great if someone could test this and give me feedback.

My two cents on this.

My initial work patches allowed to launch external applications directly from `openfortivpn` to get the session cookie, but in the end I've decided it was not the best solution. I still haven't deleted the code, you can see it here: https://github.com/gm-vm/openfortivpn/commit/b7df1a901b62e8324a2f38e1cdecab60ed44091e

I think I've mentioned my reasons already, but here a summary:
1) I could easily simulate the same behavior with a shell script and cli arguments, which helped keep `openfortivpn` simple.
2) Currently only root can run `openfortivpn`, but you really don't want to launch something like a browser as root. My initial work took care of that (and so does yours), but it turns out that if all you do is calling `setuid`, you end up in an environment without the usual variables set (XDG variables on Linux for example, but probably something else too) and that can beak some user expectations (wrong themes, wrong config paths etc...).

The idea behind your solution is not too different from mine, except for the fact that I tried to keep `openfortivpn` as independent as possible and relied solely on `stdin` and `stdout`. I see a lot of Firefox/Gecko specific code in your patch, which probably needs to be maintained as Firefox evolves. This code should probably be kept out of `openfortivpn`. I also think it only really supports Linux, while there's people using `openfortivpn` on macOS and BSD, but probably this can be fixed rather easily.

This is just my opinion as a user of `openfortivpn` and as someone who already went down this road.

**Comment by eliasaronson** [comment]
Thanks @gm-vm! With your app for getting the cookie, I have been able to log in flawlessly for months now, which can't be said for the official client which rarely worked at all.

I wrote some instructions on how I set it up and added a super simple script which automatically gets the cookie and connects to the VPN.
[https://github.com/eliasaronson/openfortivpn-cookie-script](https://github.com/eliasaronson/openfortivpn-cookie-script)

**Comment by gm-vm** [comment]
The official `openfortivpn` now has all you need, you don't have to use my fork.

I've also setup a Github action to build the Electron app straight from the repo (Linux only). I did not do it for the Qt one because then I'd have to bundle the entire Qt framework, which makes the executable really bloated.

The binaries end up here: https://github.com/gm-vm/openfortivpn-webview/releases (*WARNING*: this is a browser after all and as such it needs to be updated somewhat regularly. I clearly haven't done it in a while)

The wiki of that project also has some instructions.

**Comment by DimitriPapadopoulos** [comment]
> [...] which can't be said for the official client which rarely worked at all.

@eliasaronson By “official client”, do you mean the FortiClient downloaded from Fortinet, or `openfortivpn`? In the latter case, which version of `openfortivpn`?

**Comment by eliasaronson** [comment]
I meant FortiClient from Fortinet, which for me at least mostly never worked on Arch. Openfortivpn has always worked perfectly and now with cookie signin merged I will probably never go back to FortiClient.

**Comment by natenho** [comment]
I'm also leveraging @gm-vm solution running the electron app through @eliasaronson script, using openfortivpn 1.19.0 and it is working perfectly! It would be nice to make this app part of openfortivpn to address this scenario.

I'm just trying to figure out a good way to run it in background and without sudo (systemd?). How are you guys dealing with this? It's a little anoying to have a terminal window dedicated to keep the vpn connection up.

**Comment by anat0lius** [comment]
I'm using the `--persistent` option but it keeps disconnecting. 

How do you make it to keep alive the connection?

**Comment by DimitriPapadopoulos** [comment]
@anat0lius Hos is your question related to this issue?

**Comment by anat0lius** [comment]
> @anat0lius Hos is your question related to this issue?

Because before I had no such problems (using openfortigui), but when we were forced to use 2fa the connection started to drop if there was no traffic. 

I don't know if it worked this way even without 2fa/SAML, but I just wanted to ask and inform you.

**Comment by DimitriPapadopoulos** [comment]
If I understand you correctly, your question is totally unrelated. Please open a different ticket.

**Comment by hb0nes** [comment]
The way I've seen CLI utilities (e.g.: tsh - Teleport) supports SAML, is by having it open a browser from the commandline that navigates to the IdP, such as Okta and wait for the user.

There, the user either logs in or is already logged in, at which point the CLI continues. 

I don't know how the callback mechanism works.

**Comment by mulderr** [comment]
In new client versions there is a new option "Use external browser as user-agent for saml user authentication" which allows to well use an external browser and even reuse an existing session from said browser. (There is also another option "Enable auto-login with Azure Active Directory" but that requires the machine to be AAD joined, could be hybrid joined, which is interesting as well but a whole separate use case so lets leave that for now)

The process is described [here](https://docs.fortinet.com/document/fortigate/latest/administration-guide/364443/using-a-browser-as-an-external-user-agent-for-saml-authentication-in-an-ssl-vpn-connection).

In short FortiClient opens `localhost:8020` (by default, or whatever is configured as `saml-redirect-port`) and after successful SAML authentication external browser is redirected to `http://localhost:8020/?id=AUTHENTICATIONID`. This way FortiClient receives AUTHENTICATIONID and can request that "the SAML authentication process continues on the FortiGate with this ID".

If the whole process can be figured out in detail this puts minimal additional requirements on openfortivpn. It seems only a couple new things would be required:
* read the path to browser executable from command line or environment variable or config file
* spawn the browser with regular user privileges (drop root privileges first)
* listen on a TCP port for a GET request, parse it and read the value of parameter `id`

I think it's very much recommended for good user experience to support this use case out of the box.

I'm not quite sure of the final "cost" in complexity etc. but on first glance this looks very much doable without adding too much bloat or opinionated code. I would be willing to invest some time to experiment and maybe create a draft implementation provided that the broad concept outlined above does seem acceptable - does it?

**Comment by mulderr** [comment]
To be clear external browser in this case means, for example, vanilla Firefox/Chrome etc. (i.e. not QT webview, electronjs or Chrome DevTools Protocol shenanigans).

The way the flow is designed, after successful authentication, the browser does a callback to FC on localhost:8020 by default, with an authentication ID. Supposedly that is all that is needed to continue. I happen to currently have access to a place to test - an FG configured for SAML with Entra ID. Let me know if you'd like to pursue this - it still seems quite a bit of work to implement this correctly.

**Comment by filippor** [comment]
For external browser I implemented a script to retrieve token on repository https://github.com/filippor/XdgOpenSaml

**Comment by filippor** [comment]
see pr  #1217

**Comment by zizzencs** [comment]
Just a FYI: I created a script that can arrange the login against Azure AD without starting a browser GUI. Nice to automate the login process. The script is available here:

https://github.com/zizzencs/openfortivpn-azure-ad-login-helper

**Comment by NgoHuy** [comment]
> Just a FYI: I created a script that can arrange the login against Azure AD without starting a browser GUI. Nice to automate the login process. The script is available here:
> 
> https://github.com/zizzencs/openfortivpn-azure-ad-login-helper

[this line](https://github.com/zizzencs/openfortivpn-azure-ad-login-helper/blob/6dd06256bc751f7970a2aae978046041ef97a9b0/vpn.sh#L18) is insecure, it leaks credential on process info, you should use printf instead.

**Comment by zizzencs** [comment]
> > Just a FYI: I created a script that can arrange the login against Azure AD without starting a browser GUI. Nice to automate the login process. The script is available here:
> > https://github.com/zizzencs/openfortivpn-azure-ad-login-helper
> 
> [this line](https://github.com/zizzencs/openfortivpn-azure-ad-login-helper/blob/6dd06256bc751f7970a2aae978046041ef97a9b0/vpn.sh#L18) is insecure, it leaks credential in process info, you should use printf instead.

To be honest, security was not a priority at all here, but thanks, I'll fix it when I have some time. Passing the environment variables to the docker container might be a better solution.

**Comment by andreaippo** [comment]
> My company has just started requiring SAML. After battling a bit with the official Forticlient, which keeps dropping the connection soon after creating the tunnel, I decided to look into ways to get the VPN working with `openfortivpn`.
> 
> This is the result: https://github.com/gm-vm/openfortivpn/tree/saml
> 
> The solution is far from ideal, it requires you to login from a browser, grab the `SVPNCOOKIE` and paste it in the console, but it's good enough for me for now.
> 
> Thinking about this a bit more, I don't think there is too much room for improvement here.
> 
> One way to make things better is that of having a standardized way to launch an external application to deal with the authentication and have the `SVPNCOOKIE` passed back.
> 
> I think that supporting every SAML provider reliably in a generic way without a browser is probably close to impossible. In my case, for example, I have to use Cisco's Duo, which does more than requiring username and password.

I can't even begin to tell you how grateful I am for this!

My company recently moved to Forticlient (which has Endpoint control + VPN). On my distro, which was opensuse Tumbleweed (not officially supported), the aweful official client wasn't working. It would install, but never connect to the corporate endpoint, for whatever reason. So first pain point, try other distros and eventually distro hop to Fedora (we all know how painful and disruptive a change of distro is, with the reinstalling of everything).

On Fedora (which I had tested in a VM beforehand, of course), the client works and VPN with SAML also shows up as an option, everything fine. Except that lately I have noticed a weird trend by the client to mess up my NetworkManager wifi connection - it would leave some traces behind even after disconnecting the VPN. Namely, the DNS list would still contain corporate ones, and the domain list instead of getting blanked out, would contain `--`, which of course breaks name resolution.

That's what brought me here, I was looking for a CLI option, or better yet a NetworkManager option to be able to connect to my company's VPN without having to use Forticlient, and thanks to openfortivpn + your webview app I am now able to do just that, without any weird leftover after disconnecting (at least in my experience so far).

But better yet, since I no longer need the Forticlient UI to first connect to Endpoint control in order for the VPN SAML tab to appear in the official client, this ALSO means I can go back to opensuse Tumbleweed and just use openfortivpn + your app, and it works!

This probably also applies to all the arch etc folks out there, and it's just awesome.

TBH I ignore the consequences of staying disconnected from the "endpoint control" thingy - on Fedora the Forticlient is synchronizing with it once every minute or so - that'd be something I'd lose if I go back to opensuse. But TBH, VPN is all I need.

**Comment by andreaippo** [comment]
Update:

if you don't want to follow the wiki instructions mentioned here:
https://github.com/gm-vm/openfortivpn-webview/wiki

You may as well:

openfortivpn-webview-1.2.3-x86_64.AppImage --url <VPN_URL> | sed -n 's/^SVPNCOOKIE=\(.*\)$/\1/p' | wl-copy

This will copy the value of the cookie to your clipboard (Wayland-only, for Xorg the wl-copy command must be replaced)

This could be followed by (&&) an execution of:

sudo openfortivpn --cookie-on-stdin <VPN_host>:<VPN_port> followed by a Paste/Ctrl+Shift+V

**Comment by mrbaseman** [comment]
PR #1241 which implements saml support has been merged now on current master branch

**Comment by andreaippo** [comment]
> PR [#1241](https://github.com/adrienverge/openfortivpn/pull/1241) which implements saml support has been merged now on current master branch

Hi, and thanks so much for this!

From your changes to the `README`:

> To login using SAML you just have to open <your-vpn-domain>/remote/saml/start?redirect=1 and follow the login steps

Have you considered running `xdg-open` (and similar commands on other OSs) to open the authentication URL in the user's preferred browser? I don't know if this is considered beyond the scope of a CLI program like `openfortivpn`, but it would make the process much more seamless.

I apologize in advance if this has been already discussed/discarded, I haven't really followed the MR and all the different issues that exist on the tracker in reference to this subject.

Thanks in advance for your feedback

**Comment by exzombie** [comment]
> Have you considered running `xdg-open` (and similar commands on other OSs) to open the authentication URL in the user's preferred browser? I don't know if this is considered beyond the scope of a CLI program like `openfortivpn`, but it would make the process much more seamless.

This has been implemented in the NetworkManager plugin. See [here](https://gitlab.gnome.org/GNOME/NetworkManager-fortisslvpn/-/merge_requests/36) and [here](https://invent.kde.org/plasma/plasma-nm/-/merge_requests/378). Now that the required functionality has been merged here, these dowstream MRs can be taken out of draft status. However, I don't have time to do this right now. I would appreciate if someone is willing to test and confirm that they work with the code that was merged. I expect they work as-is, but because I implemented them against an earlier version of the SAML code, this needs to be checked.

**Comment by andreaippo** [comment]
Hi, I'm really willing to give this a try, but unfortunately I failed to build the depending projects (well, the NM plugin, I haven't even tried with plasma-nm).

`openfortivpn` built just fine following the README, but NetworkManager-fortisslvpn is really scarce on details in the README, and as I'm a Java guy I'm not really familiar with the C/Linux build process.

Here's the sequence of commands I tried after checking out your `saml-ext-browser` branch, based on https://i.sstatic.net/kuFHy.gif:

```
❯ autoscan
configure.ac: warning: missing AC_CHECK_FUNCS([memmove]) wanted by: shared/nm-utils/nm-glib.h:266
configure.ac: warning: missing AC_CHECK_FUNCS([setenv]) wanted by: src/nm-fortisslvpn-service.c:819
configure.ac: warning: missing AC_CHECK_FUNCS([setlocale]) wanted by: src/nm-fortisslvpn-service.c:767
configure.ac: warning: missing AC_CHECK_FUNCS([strchr]) wanted by: src/nm-fortisslvpn-pinentry.c:99
configure.ac: warning: missing AC_CHECK_FUNCS([strerror]) wanted by: src/nm-fortisslvpn-pppd-plugin.c:105
configure.ac: warning: missing AC_CHECK_FUNCS([strstr]) wanted by: auth-dialog/main.c:295
configure.ac: warning: missing AC_CHECK_FUNCS([strtol]) wanted by: shared/nm-fortissl-properties.c:143
configure.ac: warning: missing AC_CHECK_HEADERS([arpa/inet.h]) wanted by: properties/nm-fortisslvpn-editor-plugin.c:27
configure.ac: warning: missing AC_CHECK_HEADERS([netinet/in.h]) wanted by: properties/nm-fortisslvpn-editor-plugin.c:26
configure.ac: warning: missing AC_CHECK_HEADERS([sys/socket.h]) wanted by: src/nm-fortisslvpn-pppd-plugin.c:28
configure.ac: warning: missing AC_CHECK_HEADER_STDBOOL wanted by: shared/nm-fortissl-properties.c:33
configure.ac: warning: missing AC_C_INLINE wanted by: shared/nm-utils/gsystem-local-alloc.h:29
configure.ac: warning: missing AC_TYPE_SSIZE_T wanted by: auth-dialog/main.c:283
configure.ac: warning: missing AC_TYPE_UID_T wanted by: src/nm-fortisslvpn-pppd-plugin.c:54
```

```
❯ aclocal                   
configure.ac:124: warning: macro 'AM_GNU_GETTEXT' not found in library
configure.ac:125: warning: macro 'AM_GNU_GETTEXT_VERSION' not found in library
```

```
❯ autoheader

```

```
❯ autoconf  
configure.ac:92: warning: The macro 'AC_PROG_GCC_TRADITIONAL' is obsolete.
configure.ac:92: You should run autoupdate.
./lib/autoconf/c.m4:1676: AC_PROG_GCC_TRADITIONAL is expanded from...
configure.ac:92: the top level
```

```
❯ automake
configure.ac:11: error: required file './compile' not found
configure.ac:11:   'automake --add-missing' can install 'compile'
configure.ac:7: error: required file './install-sh' not found
configure.ac:7:   'automake --add-missing' can install 'install-sh'
configure.ac:7: error: required file './missing' not found
configure.ac:7:   'automake --add-missing' can install 'missing'
Makefile.am: error: required file './INSTALL' not found
Makefile.am:   'automake --add-missing' can install 'INSTALL'
Makefile.am:9: error: Libtool library used but 'LIBTOOL' is undefined
Makefile.am:9:   The usual way to define 'LIBTOOL' is to add 'LT_INIT'
Makefile.am:9:   to 'configure.ac' and run 'aclocal' and 'autoconf' again.
Makefile.am:9: 
Makefile.am:9:   If 'LT_INIT' is in 'configure.ac', make sure
Makefile.am:9:   its definition is in aclocal's search path.
Makefile.am:9: 
Makefile.am:9:   If you install Automake in its own prefix,
Makefile.am:9:   you'll need to arrange for the Libtool m4 files
Makefile.am:9:   to be found by aclocal.  For info on this, see:
Makefile.am:9:     https://gnu.org/s/automake/manual/automake.html#Libtool-library-used-but-LIBTOOL-is-undefined
Makefile.am: error: required file './depcomp' not found
Makefile.am:   'automake --add-missing' can install 'depcomp'
```

And that's where I gave up :(

**Comment by exzombie** [comment]
The Stackoverflow post you linked is an explanation of the build process, not instructions on what to do. Running the `autogen.sh` script in the repo should do what's needed. And then you can re-run the generated `configure` script if you want to change build settings.

However, I'm not sure where the installation prefix should be for things to work properly. I think it is distribution dependent, because NetworkManager needs to integrate quite tightly into the OS. I don't really know what the preferred setup is for developers, I just used the distro's own packages and updated them to build the test branch, which ensured that everything was where it's supposed to be. But I'm sure it's not the only way to go about this. I suggest you ask NM developers, they were pretty friendly in the IRC channel, just be patient because not everyone is in the same timezone.

**Comment by andreaippo** [comment]
> The Stackoverflow post you linked is an explanation of the build process, not instructions on what to do. Running the `autogen.sh` script in the repo should do what's needed. And then you can re-run the generated `configure` script if you want to change build settings.
> 
> However, I'm not sure where the installation prefix should be for things to work properly. I think it is distribution dependent, because NetworkManager needs to integrate quite tightly into the OS. I don't really know what the preferred setup is for developers, I just used the distro's own packages and updated them to build the test branch, which ensured that everything was where it's supposed to be. But I'm sure it's not the only way to go about this. I suggest you ask NM developers, they were pretty friendly in the IRC channel, just be patient because not everyone is in the same timezone.

Thanks. I took that route because `./autogen.sh` was failing with another error which I now forgot, so I started to doubt that running that script was the correct thing to do.

While running commands in that order I must have installed build deps that I was missing, because now that I'm back to just running `autogen` I reach a far more advanced step than before.

**Comment by Jasper-Ben** [comment]
> Have you considered running xdg-open (and similar commands on other OSs) to open the authentication URL in the user's preferred browser? I don't know if this is considered beyond the scope of a CLI program like openfortivpn, but it would make the process much more seamless.

Also interested in this. @andreaippo are you still working on it? 🙂

**Comment by andreaippo** [comment]
> > Have you considered running xdg-open (and similar commands on other OSs) to open the authentication URL in the user's preferred browser? I don't know if this is considered beyond the scope of a CLI program like openfortivpn, but it would make the process much more seamless.
> 
> Also interested in this. [@andreaippo](https://github.com/andreaippo) are you still working on it? 🙂

Hi, according to: https://github.com/adrienverge/openfortivpn/issues/867#issuecomment-2630240342, now that the openfortivpn binary supports SAML authentication, the missing pieces for providing a user friendly (read: UI) experience, are the NetworkManager plugin (which is used by multiple DEs) plus, at least in case of KDE, the NetworkManager-plasma applet.

MRs are open on both, and if you're on KDE you need to first build the NetworkManager MR, then the plasma applet MR (links in the comment I mentioned earlier).

I gave up on building those myself though, due to various issues, maybe I'll try again.

As for my current way of connecting, I'm using this inside a bash script which I place under ~/.local/bin so that I can just invoke it from a terminal anywhere:

sudo openfortivpn <your_endpoint_host>:<your_endpoint_port>/<your_auth_realm> --cookie=$(openfortivpn-webview-1.2.3-x86_64.AppImage --url https://<your_endpoint_host>:<your_endpoint_port>/<your_endpoint_realm> | sed -n 's/^SVP
       │ NCOOKIE=\(.*\)$/\1/p')

Works like a charm with minimal user interaction, except for opening a terminal, entering the alias for this script (e.g. myvpn) and entering sudo pwd if prompted. If cookie is stale you'll get a webview window/popup asking to auth, otherwise it'll reuse the current one and all you need to do is enter your sudo password.

openfortivpn-webview can be downloaded from here: https://github.com/gm-vm/openfortivpn-webview/wiki

And should of course be available on your $PATH (e.g. put it under ~/.local/bin)

**Comment by Jasper-Ben** [comment]
Hi @andreaippo,

> are the NetworkManager plugin (which is used by multiple DEs) plus, at least in case of KDE, the NetworkManager-plasma applet.
>
> MRs are open on both, and if you're on KDE you need to first build the NetworkManager MR, then the plasma applet MR (links in the comment I mentioned earlier).

That is great, but my goal to provide a solution that will work for all the Linux users in our company, regardless of their current Distribution. So even if the MRs are merged in a timely fashion and a releases are cut, I expect that it will take a considerable while until those changes have trickled down to the various distributions.

> openfortivpn-webview can be downloaded from here: https://github.com/gm-vm/openfortivpn-webview/wiki

We used openfortivpn-webview before but some users voiced security concerns about requiring such a "niche" dependency (180 LOC but whatever 🤷)

So instead my current setup involves using the nix package manager to provide users the recent 1.23.1 openfortivpn package with SAML support, regardless of the openfortivpn package version shipped by their respective distribution.

That works well, but of course it would be nicer to be able to automatically open the default browser for the authentication, rather than having to manually open the URL output from openfortivpn. So having support for xdg-open directly in openfortivpn would immediately improve the UX of our users, as I can easily provide a new version of the openfortivpn binary via nix, but since the NetworkManager and plasma applet are tightly integrated into the system I cannot simply "backport" a newer release to other distributions via nix.

> I gave up on building those myself though, due to various issues, maybe I'll try again.

No worries, far from expecting any "free labor" from you. Just wanted to know the current status. I might (that's a big might) look into it myself, if time permits. But again, this would be a "nice to have" and our current setup works well enough.

**Comment by DimitriPapadopoulos** [comment]
Because openfortivpn requires `sudo`, I think it won't be able to use the default settings of the user account.

**Comment by andreaippo** [comment]
I think I managed to build [this branch](https://gitlab.gnome.org/exzombie/NetworkManager-fortisslvpn/-/tree/saml-ext-browser) on my opensuse tumbleweed system, after installing some devel dependencies, some of them are probably not needed, but now I don't get errors when running `make` anymore:

```
- libtool
- automake
- ppp-devel
- gio-unix-2.0 is actually pkgconf
	- also installed libglibmm-2_68-1, libsigc-3_0-0 for good measure as per [[link](https://www.reddit.com/r/voidlinux/comments/he5fgl/what_should_i_do_giounix20_were_not_met/)](https://www.reddit.com/r/voidlinux/comments/he5fgl/what_should_i_do_giounix20_were_not_met/)
	- gtk3-devel + recommendations
- libsecret-devel + recommendations
- libnma-devel + recommendations
- gtk4-devel-tools
- gtk4-devel
```

I'll create a system snapshot before moving forward with `make install`

**Comment by Jasper-Ben** [comment]
> Because openfortivpn requires `sudo`, I think it won't be able to use the default settings of the user account.

Ah, that's a fair point. I was thinking alternatively an argument like `--use-saml-browser` to pass a browser binary to use for opening the link but then again that would mean that the browser would be opened with sudo privileges which wouldn't be great. So maybe the current implementation is as good as it gets without getting external tools (e.g. openfortivpn-webview involved)

**Comment by 4ndrej** [comment]
> As for my current way of connecting, I'm using this inside a bash script which I place under `~/.local/bin` so that I can just invoke it from a terminal anywhere:
> 
> ```
> sudo openfortivpn <your_endpoint_host>:<your_endpoint_port>/<your_auth_realm> --cookie=$(openfortivpn-webview-1.2.3-x86_64.AppImage --url https://<your_endpoint_host>:<your_endpoint_port>/<your_endpoint_realm> | sed -n 's/^SVPNCOOKIE=\(.*\)$/\1/p')
> ```
> 
> Works like a charm with minimal user interaction, except for opening a terminal, entering the alias for this script (e.g. myvpn) and entering sudo pwd if prompted. If cookie is stale you'll get a webview window/popup asking to auth, otherwise it'll reuse the current one and all you need to do is enter your sudo password.

As I want to share my auth session (with multifactor stuff involving mobile phone interaction) with other company services (outlook, jira, confluence, ...) as much as possible I prefer to run `https://<your_endpoint_host>:<your_endpoint_port>/<your_endpoint_realm>` in a normal permanently open Chrome tab.

I have engaged Developer tools (`F12`) in Application / cookies details there, hit `F5`, manually copy contents of SVPNCOOKIE and paste it as parameter into openfortivpn call (see below).

I also wrote short tampermonkey script listening to `sslvpn/portal.html` and `remote/login` pages that perform delayed click to move things on.

```
// @name         fortigate clicker
// @namespace    http://tampermonkey.net/
// @grant        none
// @match        https://<your_endpoint_host>/sslvpn/portal.html
// @match        https://<your_endpoint_host>/remote/login?lang=en

(function() {
    'use strict';

    var okButton = document.getElementById('ok_button');
    if (okButton!=null){
        console.log('clicking ok_button')
        window.setTimeout(function() {
            okButton.click();
        }, 1000);
    };

    var samlButton = document.getElementById('saml-login-bn');
    if (samlButton!=null){
        console.log('clicking samlButton')
        window.setTimeout(function() {
            samlButton.click();
        }, 1000);
    };

})();
```

My vpn script looks like this:
```
read -p "paste SVNPCOOKIE here please: " SVPNCOOKIE
su -c "openfortivpn <your_endpoint_host>:<your_endpoint_port> --cookie=${SVPNCOOKIE}" - root
```

This works very well, but YMMW.

**Comment by andreaippo** [comment]
I followed the KDE build guide here:

https://develop.kde.org/docs/getting-started/building/kde-builder-setup/

Finally managed to build plasma-nm (probably not with the NetworkManager build from my previous step, but that doesn't matter right now). I don't know at this point how to run the plasma-nm I just built.

```
.
├── build
│   ├── breeze-icons
│   ├── extra-cmake-modules
│   ├── gpgme
│   ├── karchive
│   ├── kauth
│   ├── kbookmarks
│   ├── kcalc
│   ├── kcmutils
│   ├── kcodecs
│   ├── kcolorscheme
│   ├── kcompletion
│   ├── kconfig
│   ├── kconfigwidgets
│   ├── kcoreaddons
│   ├── kcrash
│   ├── kdbusaddons
│   ├── kdoctools
│   ├── kglobalaccel
│   ├── kguiaddons
│   ├── ki18n
│   ├── kiconthemes
│   ├── kio
│   ├── kirigami
│   ├── kitemviews
│   ├── kjobwidgets
│   ├── knotifications
│   ├── kpackage
│   ├── kparts
│   ├── kpipewire
│   ├── kservice
│   ├── ksvg
│   ├── ktextwidgets
│   ├── kwallet
│   ├── kwayland
│   ├── kwidgetsaddons
│   ├── kwindowsystem
│   ├── kxmlgui
│   ├── libplasma
│   ├── modemmanager-qt
│   ├── networkmanager-qt
│   ├── plasma-activities
│   ├── plasma-nm
│   ├── plasma-wayland-protocols
│   ├── polkit-qt-1
│   ├── qca
│   ├── qcoro
│   ├── selenium-webdriver-at-spi
│   ├── solid
│   ├── sonnet
│   ├── wayland
│   └── wayland-protocols
├── log
│   ├── 2025-02-21_01
│   ├── 2025-02-21_03
│   ├── 2025-02-21_04
│   ├── latest
│   └── latest-by-phase
├── src
│   ├── breeze-icons
│   ├── extra-cmake-modules
│   ├── gpgme
│   ├── karchive
│   ├── kauth
│   ├── kbookmarks
│   ├── kcalc
│   ├── kcmutils
│   ├── kcodecs
│   ├── kcolorscheme
│   ├── kcompletion
│   ├── kconfig
│   ├── kconfigwidgets
│   ├── kcoreaddons
│   ├── kcrash
│   ├── kdbusaddons
│   ├── kdoctools
│   ├── kglobalaccel
│   ├── kguiaddons
│   ├── ki18n
│   ├── kiconthemes
│   ├── kio
│   ├── kirigami
│   ├── kitemviews
│   ├── kjobwidgets
│   ├── knotifications
│   ├── kpackage
│   ├── kparts
│   ├── kpipewire
│   ├── kservice
│   ├── ksvg
│   ├── ktextwidgets
│   ├── kwallet
│   ├── kwayland
│   ├── kwidgetsaddons
│   ├── kwindowsystem
│   ├── kxmlgui
│   ├── libplasma
│   ├── modemmanager-qt
│   ├── networkmanager-qt
│   ├── plasma-activities
│   ├── plasma-nm
│   ├── plasma-wayland-protocols
│   ├── polkit-qt-1
│   ├── qca
│   ├── qcoro
│   ├── selenium-webdriver-at-spi
│   ├── solid
│   ├── sonnet
│   ├── wayland
│   └── wayland-protocols
└── usr
    ├── bin
    ├── include
    ├── lib
    ├── lib64
    ├── mkspecs
    └── share

118 directories, 0 files
```

**Comment by exzombie** [comment]
@andreaippo A Plasma applet needs to be installed in order for Plasma to find it. Have you tried using `plasmapkg2`?

I'm not a "regular" NM or plasma developer, so I don't know what a smooth development setup looks like. For my purposes, I was happy with pulling the source packages for my distro (which is Ubuntu), replacing the sources, then building and installing the deb packages. That solves all the integration issues, but for plasma-nm, I had to cherry-pick commits because Ubuntu 22.04 packages a much older plasma-nm, for Plasma 5. I'd venture to guess that's something you don't want to deal with because most people don't, but if you're not "most people", it will produce a working package that you can then simply keep using until everything is upstream in (hopefully) the next version of your distro.
